File: Graph/problem133/problem133.py
from typing import Optional


class Node:
    def __init__(self, val: int = 0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


class Solution:
    def cloneGraph(self, node: Optional[Node]) -> Optional[Node]:
        if not node:
            return None

        visited: dict[Node, Node] = {}

        def dfs(n: Node) -> Node:
            if n in visited:
                return visited[n]
            clone = Node(n.val)
            visited[n] = clone
            for nb in n.neighbors:
                clone.neighbors.append(dfs(nb))
            return clone

        return dfs(node)

File: Graph/problem133/problem133.cpp
#include <unordered_map>
#include <vector>

using namespace std;


#ifndef ONLINE_JUDGE
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() : val(0) {}
    Node(int v) : val(v) {}
    Node(int v, vector<Node*> nb) : val(v), neighbors(nb) {}
};
#endif

class Solution {
    unordered_map<Node*, Node*> visited;

    Node* dfs(Node* node) {
        if (!node) {
            return nullptr;
        }
        auto it = visited.find(node);
        if (it != visited.end()) {
            return it->second;
        }
        Node* clone = new Node(node->val);
        visited[node] = clone;
        for (Node* nb : node->neighbors) {
            clone->neighbors.push_back(dfs(nb));
        }
        return clone;
    }

public:
    Node* cloneGraph(Node* node) {
        visited.clear();
        return dfs(node);
    }
};
